### ISSUE TITLE
knife ec backup with purge does not cleanup ACLs for clients and nodes

### ISSUE THREAD

**Issue by PrajaktaPurohit** [chef/chef-server#2842]
When running knife ec backup with the purge option, acl files are not cleaned up during the purge process for deleted clients and nodes. This makes manual intervention required to cleanup the backups and prevents having automated disaster-recovery replication in place using knife ec backup to an alternate location as node and client deletions are normal events and without the ACL entries being purged errors are generated during the restore for these orphaned objects.

https://github.com/chef/knife-ec-backup/issues/162

Acceptance criteria:
- [ ] Update Readme in knife-ec-backup to include info: --purge can be used only to sync users. (would --sync_users be a better choice of words?) : This is not true, knife-ec-backup can be used to purge all of the chef objects.
  https://github.com/chef/knife-ec-backup/pull/170 - should be closed.
  ```
  --purge: Whether to sync user deletions from backup source to restore destination. (default: false)
  This can only be used to sync **_user_** deletions.
  ```
- [ ] Offer potential workaround. - The fix should be in knife ec backup.
       Rename (`mv`) old backup directory, create new backup using `knife ec backup`.  Does this accomplish the same thing?
       This could be placed in a script in order to automate the transaction.  See demonstration below.
- [x] Client and node ACLs are deleted from a backup directory when `--purge` option is given.
      https://github.com/chef/chef/pull/12458
      https://github.com/chef/knife-ec-backup/pull/171
- [x] What is server doing?
        - Does server delete the acl for the object that is deleted - yes it does.
- [x] What is knife ec backup doing?
        - Does knife ec backup fetch the acls only for the objects that exist? - This PR should fix the issue with knife ec backup.
- [x] What is chef_fs in client doing?
       - the purge option gets included into config (probably via mixlib). This is also sent to chef_fs. With all options and recurse_depth of nil. - yes chef_fs is also doing the right thing in this case.
       - knife ec backup manages users and orgs separately and the rest of the paths are per : https://github.com/chef/knife-ec-backup/blob/main/lib/chef/knife/ec_backup.rb#L180
       - https://github.com/chef/chef/blob/main/lib/chef/chef_fs/file_system.rb#L326 seems to indicate that nil recurse_depth implies recurse indefinitely (acl dir has 2 levels of dir)

See:
https://github.com/chef/knife-ec-backup/pull/170
https://github.com/chef/chef/pull/12458

Acceptance Criteria:
- [ ] knife-ec-backup purge works as expected with the latest chef-server.
- [ ] backup with chef-server 12.17.15 and restore on latest chef-server (14.12) with the patch from the PS should work fine.
- [ ] Documentation to define what versions this does not work with (eg Server 11) and the upgrade path from that version.
- [ ] Issue to define the umbrella test to test knife-ec backup --purge testing on one of the scenarios.
- [ ] Update knife-ec-backup spec

**Comment by lbakerchef** [comment]
```
#
# Reproduce issue
#

# Create some data
[root@ip-10-0-9-56 ~]# knife node create node1
Created node[node1]
[root@ip-10-0-9-56 ~]# knife client create client1
Created client[client1]
[root@ip-10-0-9-56 ~]# knife node list
node1
[root@ip-10-0-9-56 ~]# knife client list
4thcoffee-validator
client1

# Take an initial backup using knife ec backup
[root@ip-10-0-9-56 ~]# knife ec backup /backup --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-9-56.us-west-2.compute.internal

# Delete a node and/or client from the Chef Infra Server
[root@ip-10-0-9-56 ~]# knife node delete -y node1
Deleted node[node1]
[root@ip-10-0-9-56 ~]# knife client delete -y client1
Deleted client[client1]
[root@ip-10-0-9-56 ~]# knife node list

[root@ip-10-0-9-56 ~]# knife client list
4thcoffee-validator

# Run knife ec backup again with the --purge option
[root@ip-10-0-9-56 ~]# knife ec backup /backup --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-9-56.us-west-2.compute.internal

# Find ACLs
[root@ip-10-0-9-56 ~]# find /backup/ -name node1.json
/backup/organizations/4thcoffee/acls/nodes/node1.json
[root@ip-10-0-9-56 ~]# find /backup/ -name client1.json
/backup/organizations/4thcoffee/acls/clients/client1.json
```

**Comment by lbakerchef** [comment]
Observation:  The problem exists only for backing up to the SAME directory.  If you backup to a DIFFERENT directory, the problem goes away.  This seems to make sense.  If you want to preserve data, this is working as intended, so just save off the old backup, and make a new one.  If you don't want to preserve data, erase the backup dir before making new backup.  But it depends on the purpose of the `--purge` option.
```
[root@ip-10-0-9-56 ~]# knife ec backup /backup1 --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-9-56.us-west-2.compute.internal
[root@ip-10-0-9-56 ~]# find /backup1/ -name client1.json
[root@ip-10-0-9-56 ~]# find /backup1/ -name node1.json
```
Question:  Is `--purge` supposed to delete files in a backup directory?

**Comment by lbakerchef** [comment]
What's the difference if you don't use the `--purge` option?
```
# what happens without purge?

[root@ip-10-0-9-56 ~]# knife node list
node1
node2
[root@ip-10-0-9-56 ~]# knife client list
4thcoffee-validator
client1
client2
[root@ip-10-0-9-56 ~]# knife ec backup /backup-predelete --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-9-56.us-west-2.compute.internal
[root@ip-10-0-9-56 ~]# find /backup-predelete/ -name node1.json
/backup-predelete/organizations/4thcoffee/acls/nodes/node1.json
/backup-predelete/organizations/4thcoffee/nodes/node1.json
[root@ip-10-0-9-56 ~]# find /backup-predelete/ -name node2.json
/backup-predelete/organizations/4thcoffee/acls/nodes/node2.json
/backup-predelete/organizations/4thcoffee/nodes/node2.json
[root@ip-10-0-9-56 ~]# find /backup-predelete/ -name client1.json
/backup-predelete/organizations/4thcoffee/acls/clients/client1.json
/backup-predelete/organizations/4thcoffee/clients/client1.json
[root@ip-10-0-9-56 ~]# find /backup-predelete/ -name client2.json
/backup-predelete/organizations/4thcoffee/acls/clients/client2.json
/backup-predelete/organizations/4thcoffee/clients/client2.json
[root@ip-10-0-9-56 ~]# knife node delete node1
Do you really want to delete node1? (Y/N) Y
Deleted node[node1]
[root@ip-10-0-9-56 ~]# knife node delete node2 -y
Deleted node[node2]
[root@ip-10-0-9-56 ~]# knife client delete client1 -y
Deleted client[client1]
[root@ip-10-0-9-56 ~]# knife client delete client2 -y
Deleted client[client2]
[root@ip-10-0-9-56 ~]# knife node list

[root@ip-10-0-9-56 ~]# knife client list
4thcoffee-validator
[root@ip-10-0-9-56 ~]# knife ec backup /backup-postdelete-nopurge --sql-db automate-cs-oc-erchef -V -s https://ip-10-0-9-56.us-west-2.compute.internal
[root@ip-10-0-9-56 ~]# find /backup-postdelete-nopurge/ -name node1.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-nopurge/ -name node2.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-nopurge/ -name client1.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-nopurge/ -name client2.json
[root@ip-10-0-9-56 ~]# knife ec backup /backup-postdelete-purge --sql-db automate-cs-oc-erchef --purge -V -s https://ip-10-0-9-56.us-west-2.compute.internal
[root@ip-10-0-9-56 ~]# find /backup-postdelete-purge/ -name node1.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-purge/ -name node2.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-purge/ -name client1.json
[root@ip-10-0-9-56 ~]# find /backup-postdelete-purge/ -name client2.json
[root@ip-10-0-9-56 ~]# diff --brief --recursive /backup-postdelete-nopurge /backup-postdelete-purge
```

**Comment by lbakerchef** [comment]
The problem is sporadic and intermittent.  Today, I can't reproduce the issue.

**Comment by lbakerchef** [comment]
Question:  @collinmcneese @PrajaktaPurohit Does anyone know if this worked in the past, and then stopped working?  Or was this functionality just never there?  I'm trying to find out if I should be looking for existing broken code to fix, or if I should be looking to write new functionality that was never there to begin with.

**Comment by lbakerchef** [comment]
The `--purge` option can only be used to sync users.  It works when used in this capacity.

```
root@ip-10-0-4-202:~# chef-server-ctl user-list
jackdoe
janedoe
jilldoe
johndoe
johndoex
pivotal

# Take an initial backup using knife ec backup
root@ip-10-0-4-202:~# knife ec backup /backup --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-4-202.us-west-2.compute.internal
# snip
root@ip-10-0-4-202:~# cp -r /backup /backup.BAK

# Delete users
root@ip-10-0-4-202:~# chef-server-ctl user-delete johndoe
Do you want to delete the user johndoe? (Y/N) y
Checking organization memberships...
Deleting user johndoe.
root@ip-10-0-4-202:~# chef-server-ctl user-delete johndoex
Do you want to delete the user johndoex? (Y/N) y
Checking organization memberships...
Checking admin group memberships for 1 org(s).
Deleting user johndoex.
root@ip-10-0-4-202:~# chef-server-ctl user-delete jackdoe
Do you want to delete the user jackdoe? (Y/N) y
Checking organization memberships...
Deleting user jackdoe.
root@ip-10-0-4-202:~# chef-server-ctl user-delete jilldoe
Do you want to delete the user jilldoe? (Y/N) y
Checking organization memberships...
Deleting user jilldoe.
root@ip-10-0-4-202:~# chef-server-ctl user-list
janedoe
pivotal

# Take another backup with the --purge option
root@ip-10-0-4-202:~# knife ec backup /backup --sql-db automate-cs-oc-erchef  --purge -V -s https://ip-10-0-4-202.us-west-2.compute.internal
Downloading Users
Deleting user jackdoe from local backup (purge is on)
Deleting user johndoex from local backup (purge is on)
Deleting user jilldoe from local backup (purge is on)
Deleting user johndoe from local backup (purge is on)
Downloading organization object for 4thcoffee from https://ip-10-0-4-202.us-west-2.compute.internal/organizations/4thcoffee
Copying /acls/groups/billing-admins.json
Copying /groups/billing-admins.json
Copying /groups/admins.json
Copying /acls/groups/public_key_read_access.json
Copying /groups/public_key_read_access.json
Copying /clients
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /invitations.json
Updated /invitations.json
Copying /members.json
Updated /members.json
Copying /nodes
Copying /org.json
Copying /policies
Copying /policy_groups
Copying /roles
Copying /acls/organization.json
Copying /acls/clients/4thcoffee-validator.json
Copying /acls/containers/clients.json
Copying /acls/containers/containers.json
Copying /acls/containers/cookbook_artifacts.json
Copying /acls/containers/cookbooks.json
Copying /acls/containers/data.json
Copying /acls/containers/environments.json
Copying /acls/containers/groups.json
Copying /acls/containers/nodes.json
Copying /acls/containers/policies.json
Copying /acls/containers/policy_groups.json
Copying /acls/containers/roles.json
Copying /acls/containers/sandboxes.json
Copying /acls/environments/_default.json
Copying /acls/groups/000000000000ec9aba30060322d83390.json
Copying /acls/groups/admins.json
Copying /acls/groups/clients.json
Copying /acls/groups/users.json
Copying /groups/000000000000ec9aba30060322d83390.json
Copying /groups/admins.json
Copying /groups/clients.json
Copying /groups/users.json
Updated /groups/users.json
** Finished **

# Find ACLs, compare with original backup
root@ip-10-0-4-202:~# find /backup.BAK/users -name *.json
/backup.BAK/users/jackdoe.json
/backup.BAK/users/johndoex.json
/backup.BAK/users/pivotal.json
/backup.BAK/users/jilldoe.json
/backup.BAK/users/janedoe.json
/backup.BAK/users/johndoe.json
root@ip-10-0-4-202:~# find /backup.BAK/user_acls/ -name *.json
/backup.BAK/user_acls/jackdoe.json
/backup.BAK/user_acls/johndoex.json
/backup.BAK/user_acls/pivotal.json
/backup.BAK/user_acls/jilldoe.json
/backup.BAK/user_acls/janedoe.json
/backup.BAK/user_acls/johndoe.json
root@ip-10-0-4-202:~# find /backup/users -name *.json
/backup/users/pivotal.json
/backup/users/janedoe.json
root@ip-10-0-4-202:~# find /backup/user_acls/ -name *.json
/backup/user_acls/pivotal.json
/backup/user_acls/janedoe.json
```

**Comment by lbakerchef** [comment]
Relevant Slack comments:
```
Prajakta Purohit
Couple of question wrt to this change, would it break with the older versions of Chefserver?
Do we care if it fails from 11 -> 14? There is not a direct migration path there anyways (edited) 

Marc Paradise
That is the next thing to explore there. I think we'd expecit it to fail on 11, but not after 12 or 12.something.

Tim Smith
11 breaking is fine

Prajakta Purohit
so a sufficient test would be that 12 onwards we are able to backup --purge and restore with knife-ec-backup? (edited) 

Marc Paradise
+ verifying that all acls match up
```

**Comment by PrajaktaPurohit** [comment]
This fix is available on Chef Infra Server 14.12.21
